### PR TITLE
Plane: Fix inability to climb the last few meters of takeoff

### PR DESCRIPTION
**Pull request by Georacer**
## Summary

This PR addresses #28685.

## Details

This affects takeoffs without an airspeed sensor.

### Cause

Some airframes will not climb at 0 pitch, regardless of throttle setting.
Thus, the current takeoff level-off strategy of asymptotically forcing pitch to 0, results in the aircraft to never climb the final few meters.

### Reproduction

The behaviour was reproduced in SITL, by setting:
- `PTCH_TRIM_DEG`: -10.0
- `TKOFF_THR_MAX`: 75.0

### Solution

After the expected time to complete the takeoff has elapsed, drive the desired pitch gradually back to takeoff pitch, over the next 5 seconds.

### Testing

**Before**:
image

**After**:
image

### Alternatives

The proposed solution of widening the TECS pitch limits and letting TECS take over and regulate the last few meters of the climb was not feasible:
During an airspeedless takeoff, TECS uses throttle to correct for altitude errors.
It would use pitch (if it was allowed) to correct for airspeed errors).
Thus letting TECS command the pitch angle would not help.

### CODE REVIEW

**Comment by menschel** [comment]
Weather is currently not favorable for a test flight but I'll try to test in case it gets better.

**Comment by menschel** [comment]
Two short flights.
[24-11-23_08-37-52.bin](https://github.com/menschel/ardupilot/blob/stash/build/bin/Arduplane/24-11-23_08-37-52.bin)
[24-11-23_08-42-26.bin](https://github.com/menschel/ardupilot/blob/stash/build/bin/Arduplane/24-11-23_08-42-26.bin)
Fun fact, the overshoot now happens based on the workaround.

IMO the whole thing is connected to a false climb-rate estimation. I have logs from a landing test session with 4.5.7 and I see level-off happening at 3m below target altitude, not 8m.
It may be a good idea to stop limiting the pitch at level-off and see what TECS is doing out of it. If I'm not mistaken, the waypoint is 1km away and TECS should linearly climb.

File: ArduPlane/takeoff.cpp [hunk 1]
```diff
@@ -304,6 +304,14 @@ int16_t Plane::get_takeoff_pitch_min_cd(void)
         // if height-below-target has been initialized then use it to create and apply a scaler to the pitch_min
         if (auto_state.height_below_takeoff_to_level_off_cm != 0) {
             float scalar = remaining_height_to_target_cm / (float)auto_state.height_below_takeoff_to_level_off_cm;
+
+            const int32_t takeoff_end_delay_ms = AP_HAL::millis() - takeoff_state.level_off_start_time_ms - g.takeoff_pitch_limit_reduction_sec*1000;
```
Comment (peterbarker) [inline ArduPlane/takeoff.cpp:308]:
```suggestion
            const uint32_t takeoff_end_delay_ms = AP_HAL::millis() - takeoff_state.level_off_start_time_ms - g.takeoff_pitch_limit_reduction_sec*1000;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
@Georacer I'm surprised you didn't implement the expanding pitch window using TECS like we discussed, did it not work out?

**Comment by Georacer** [comment]
> @Georacer I'm surprised you didn't implement the expanding pitch window using TECS like we discussed, did it not work out?

As I wrote in the description, the proposed solution of widening the TECS pitch limits and letting TECS take over and regulate the last few meters of the climb was not feasible:
During an airspeedless takeoff, TECS uses throttle to correct for altitude errors.
It would use pitch (if it was allowed) to correct for airspeed errors; this is not reliable in an airspeedless case, which we are in.
Thus letting TECS command the pitch angle would not help.

I did try it and it didn't give effective results.
If you're really curious I can repeat the experiment.

File: ArduPlane/takeoff.cpp [hunk 2]
```diff
@@ -304,6 +304,14 @@ int16_t Plane::get_takeoff_pitch_min_cd(void)
...
+            if (takeoff_end_delay_ms > 0) {
```
Comment (peterbarker) [inline ArduPlane/takeoff.cpp:309]: This might be a `linear_interpolate`?

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
discussed on EU dev call, decided just to declare the takeoff complete if we are in the taper off region and we timeout

**Comment by menschel** [comment]
Can we also do something about the now forced 100% throttle?

I argue that it's a wrong design decision not to fall back to TECS controlled throttle here.
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/takeoff.cpp#L282

It also adds to the issue here because it influences the climb-rate estimation.

**Comment by Georacer** [comment]
> Can we also do something about the now forced 100% throttle?
> 
> I argue that it's a wrong design decision not to fall back to TECS controlled throttle, means the airspeed sensor presence should not be evaluated here. https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/takeoff.cpp#L282
> 
> It also adds to the issue here because it influences the climb-rate estimation.

For this PR we definitely won't change the throttle strategy.

Having a constant THR_MAX was a request from other developers, so you'd have to make your case on why this is a good idea.
Preferably by opening a new ticket.

**Comment by menschel** [comment]
I've been spinning the numbers, with level off at 42m, distance 8m, climb_rate 4mps, I came up with this calculation that results in the observed behavior.

```
dh = climb_rate * t

climb_rate = (pitch/pitch_max) * climb_max

pitch = pitch_initial * (takeoff_pitch_limit_reduction_sec - t)

dh = (((pitch_initial * (takeoff_pitch_limit_reduction_sec - t))  /pitch_max) * climb_max) * t

dh = (((12 * (2 - t))  /18) * 3) * t

dh = 3t - 2*t²

Integral(dh) = (3/2)t² - (2/3)t³
from 0 to takeoff_pitch_limit_reduction_sec = 2

(3/2)takeoff_pitch_limit_reduction_sec² - (2/3)takeoff_pitch_limit_reduction_sec³

(3/2)*4 - 2/3*8

6-(1/12) = 5.916

--> 47.916m
```

I'm still working on the math but maybe there is some sqrt(g.takeoff_pitch_limit_reduction_sec)) missing, maybe here.

```
float sec_to_target = (remaining_height_to_target_cm * 0.01f * sqrt(g.takeoff_pitch_limit_reduction_sec)) / (-auto_state.sink_rate);
```

**Comment by menschel** [comment]
I have created alternative PR https://github.com/ArduPilot/ardupilot/pull/28776
Mea Culpa, Mea Culpa.

**Comment by menschel** [comment]
I believe I know what it is

`float sec_to_target = (remaining_height_to_target_cm * 0.01f) / (-auto_state.sink_rate);`

the interval i.e. the altitude gained is only half in what that time because of the linear decreasing pitch and consequently climb-rate. 

`float sec_to_target = (remaining_height_to_target_cm * 0.01f * 2) / (-auto_state.sink_rate);`
is the correct time to target.

If I may do so I'll include your commit d698139f0c6dfe3ce965d722b8c73be607de6d44 into my [PR](https://github.com/ArduPilot/ardupilot/pull/28776)

**Comment by Georacer** [comment]
Superseded by #28792

**Pull request closed without merging**